src/format/Kdbx3Reader.cpp

```diff
@@ -0,0 +1,376 @@
+    if (ok && signature2 == KeePass1::SIGNATURE_2) {
+        raiseError(tr("The selected file is an old KeePass 1 database (.kdb).\n\n"
+                          "You can import it by clicking on Database > 'Import KeePass 1 database...'.\n"
+                          "This is a one-way migration. You won't be able to open the imported "
+                          "database with the old KeePassX 0.4 version."));
+        return nullptr;
+    } else if (!ok || signature2 != KeePass2::SIGNATURE_2) {
+        raiseError(tr("Not a KeePass database."));
+        return nullptr;
+    }
+
+    quint32 version = Endian::readSizedInt<quint32>(m_headerStream, KeePass2::BYTEORDER, &ok)
+        & KeePass2::FILE_VERSION_CRITICAL_MASK;
+    quint32 maxVersion = KeePass2::FILE_VERSION & KeePass2::FILE_VERSION_CRITICAL_MASK;
+    if (!ok || (version < KeePass2::FILE_VERSION_MIN) || (version > maxVersion)) {
+        raiseError(tr("Unsupported KeePass KDBX 2 or 3 database version."));
+        return nullptr;
+    }
+
+    while (readHeaderField() && !hasError()) {
+    }
+
+    headerStream.close();
+
+    if (hasError()) {
+        return nullptr;
+    }
+
+    // check if all required headers were present
+    if (m_masterSeed.isEmpty() || m_encryptionIV.isEmpty()
+        || m_streamStartBytes.isEmpty() || m_protectedStreamKey.isEmpty()
+        || m_db->cipher().isNull()) {
+        raiseError("missing database headers");
+        return nullptr;
+    }
+
+    if (!m_db->setKey(key, false)) {
+        raiseError(tr("Unable to calculate master key"));
+        return nullptr;
+    }
+
+    if (m_db->challengeMasterSeed(m_masterSeed) == false) {
+        raiseError(tr("Unable to issue challenge-response."));
+        return nullptr;
+    }
+
+    CryptoHash hash(CryptoHash::Sha256);
+    hash.addData(m_masterSeed);
+    hash.addData(m_db->challengeResponseKey());
+    hash.addData(m_db->transformedMasterKey());
+    QByteArray finalKey = hash.result();
+
+    SymmetricCipher::Algorithm cipher = SymmetricCipher::cipherToAlgorithm(m_db->cipher());
+    SymmetricCipherStream cipherStream(m_device, cipher,
+                                       SymmetricCipher::algorithmMode(cipher), SymmetricCipher::Decrypt);
+    if (!cipherStream.init(finalKey, m_encryptionIV)) {
+        raiseError(cipherStream.errorString());
+        return nullptr;
+    }
+    if (!cipherStream.open(QIODevice::ReadOnly)) {
+        raiseError(cipherStream.errorString());
+        return nullptr;
+    }
+
+    QByteArray realStart = cipherStream.read(32);
+
+    if (realStart != m_streamStartBytes) {
+        raiseError(tr("Wrong key or database file is corrupt."));
+        return nullptr;
+    }
+
+    HashedBlockStream hashedStream(&cipherStream);
+    if (!hashedStream.open(QIODevice::ReadOnly)) {
+        raiseError(hashedStream.errorString());
+        return nullptr;
+    }
+
+    QIODevice* xmlDevice;
+    QScopedPointer<QtIOCompressor> ioCompressor;
```

Comment:
What do you think of replacing `QtIOCompressor` with `KCompressionDevice` from KF5Archive?

Reply:
What would be the purpose?

Reply:
Remove the embedded version.

Reply:
I'm against adding new dependencies here if not really needed.

Reply:
Ok; makes sense. I've also noticed there is no package available for Ubuntu Trusty. Please disregard.